File: ruby/clock/clock.rb
class Clock
  attr_reader :minutes

  def initialize(hour: 0, minute: 0)
    @minutes = (hour * MINUTES_PER_HOUR + minute).modulo(MINUTES_PER_DAY)
  end 

  def +(other) = Clock.new(minute: minutes + other.minutes)
  def -(other) = Clock.new(minute: minutes - other.minutes)
  def ==(other) = other.class == self.class && other.minutes == minutes
  def to_s = format('%02d:%02d', *minutes.divmod(MINUTES_PER_HOUR))

  HOURS_PER_DAY = 24
  MINUTES_PER_HOUR = 60
  MINUTES_PER_DAY = MINUTES_PER_HOUR * HOURS_PER_DAY
  private_constant :HOURS_PER_DAY, :MINUTES_PER_HOUR, :MINUTES_PER_DAY
end

File: elixir/clock/lib/clock.ex
defmodule Clock do
  defstruct hour: 0, minute: 0

  @minutes_per_day 1440
  @minutes_per_hour 60

  def new(hour, minute) do
    minutes_in_day = Integer.mod(hour * @minutes_per_hour + minute, @minutes_per_day)

    %Clock{
      hour: div(minutes_in_day, @minutes_per_hour),
      minute: rem(minutes_in_day, @minutes_per_hour)
    }
  end

  def add(%Clock{hour: hour, minute: minute}, add_minute) do
    new(hour, minute + add_minute)
  end

  defimpl String.Chars do
    def to_string(%Clock{hour: hour, minute: minute}) do
      :io_lib.format("~2..0w:~2..0w", [hour, minute]) |> Kernel.to_string()
    end
  end
end
